### PR TITLE
[MPS] Fix unique flatten logic

### LINKED ISSUE #104879: `Tensor.unique` bug on MPS
### 🐛 Describe the bug

The output of

```python
import torch

a = torch.tensor([[[0, 1]]])
print(a.unique())
print(a.to("mps").unique())
print(a.to("mps").cpu().unique())
```

is
```
tensor([0, 1])
tensor([0], device='mps:0')
tensor([0, 1])
```

interestingly, if `a = torch.tensor([0, 1])` then it works as expected.

### Versions

Collecting environment information...
PyTorch version: 2.0.1
Is debug build: False
CUDA used to build PyTorch: None
ROCM used to build PyTorch: N/A

OS: macOS 13.4.1 (arm64)
GCC version: Could not collect
Clang version: 14.0.3 (clang-1403.0.22.14.1)
CMake version: version 3.26.4
Libc version: N/A

Python version: 3.10.11 (main, Apr 20 2023, 13:58:42) [Clang 14.0.6 ] (64-bit runtime)
Python platform: macOS-13.4.1-arm64-arm-64bit
Is CUDA available: False
CUDA runtime version: No CUDA
CUDA_MODULE_LOADING set to: N/A
GPU models and configuration: No CUDA
Nvidia driver version: No CUDA
cuDNN version: No CUDA
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

CPU:
Apple M2 Pro

Versions of relevant libraries:
[pip3] numpy==1.24.3
[pip3] torch==2.0.1
[pip3] torchvision==0.15.2
[conda] numpy                     1.24.3                   pypi_0    pypi
[conda] torch                     2.0.1                    pypi_0    pypi
[conda] torchvision               0.15.2                   pypi_0    pypi

cc @kulinseth @albanD @malfet @DenisVieriu97 @razarmehr @abhudev

### PR DESCRIPTION
**Pull request by malfet**
Tensor must be flatted if dim is none before checking whether or not dim dimension is already None

Fixes https://github.com/pytorch/pytorch/issues/104879

### CODE REVIEW

**Review by albanD — COMMENTED** [review]
This is just code re-ordering right?

**Review by albanD — APPROVED** [review]
Sounds good

**Comment by malfet** [comment]
> This is just code re-ordering right?

Yes, reordering + adding a regression test
